packages/style-value-parser/src/at-queries-from-tokens/media-query.js

```diff
@@ -0,0 +1,241 @@
+      : this.value.toString();
+    return `(${this.key}: ${value})`;
+  }
+  static get parser(): TokenParser<MediaQuerySinglePair> {
+    return TokenParser.sequence(
+      TokenParser.tokens.OpenParen,
+      TokenParser.tokens.Ident.map((token) => token[4].value, '.stringValue'),
+      TokenParser.tokens.Colon,
+      TokenParser.oneOf(
+        TokenParser.tokens.Dimension.map(
+          (token) => `${token[4].value}${token[4].unit}`,
+        ),
+        TokenParser.tokens.Ident.map((token) => token[4].value),
+        TokenParser.sequence(
+          TokenParser.tokens.Number.map((token) => token[4].value),
+          TokenParser.tokens.Delim.where(
+            (token): implies token is TokenDelim => token[4].value === '/',
+          ).map(() => '/'),
+          TokenParser.tokens.Number.map((token) => token[4].value),
+        ).separatedBy(TokenParser.tokens.Whitespace.optional),
+      ),
+      TokenParser.tokens.CloseParen,
+    )
+      .separatedBy(TokenParser.tokens.Whitespace.optional)
+      .map(
+        ([_a, key, _c, value, _d, _e]) => new MediaQuerySinglePair(key, value),
+      );
+  }
+}
+
+export class NotMediaRule {
+  rule: MediaQuerySinglePair | MediaQuerySingleWordCondition;
+  constructor(rule: this['rule']) {
+    this.rule = rule;
+  }
+  toString(): string {
+    return `(not ${this.rule.toString()})`;
+  }
+  static get parser(): TokenParser<NotMediaRule> {
+    return TokenParser.sequence(
+      TokenParser.tokens.OpenParen,
+      TokenParser.tokens.Ident.map(
+        (token) => token[4].value,
+        '.stringValue',
+      ).where((key) => key === 'not'),
+      TokenParser.tokens.Whitespace,
+      TokenParser.oneOf(
+        MediaQuerySinglePair.parser,
+        MediaQuerySingleWordCondition.parser,
+      ),
+      TokenParser.tokens.CloseParen,
+    ).map(([_open, _not, _space, rule, _close]) => new NotMediaRule(rule));
+  }
+}
+
+export class AndSeparatedMediaRules {
+  queries: $ReadOnlyArray<MediaQueryKeywords | MediaRule | NotMediaRule>;
+  constructor(queries: this['queries']) {
+    this.queries = queries;
+  }
+  toString(): string {
+    return this.queries.map((query) => query.toString()).join(' and ');
+  }
+  static get parser(): TokenParser<AndSeparatedMediaRules> {
+    return TokenParser.oneOrMore(
+      TokenParser.oneOf(
+        MediaQueryKeywords.parser,
+        // () => OrSeparatedMediaRules.parser,
+        NotMediaRule.parser,
+        MediaRule.parser,
+      ),
+    )
+      .separatedBy(
+        TokenParser.string('and').surroundedBy(TokenParser.tokens.Whitespace),
+      )
+      .map((rules) => new AndSeparatedMediaRules(rules));
+  }
+}
+
+export class OrSeparatedMediaRules {
+  queries: $ReadOnlyArray<
+    MediaQueryKeywords | MediaRule | NotMediaRule | AndSeparatedMediaRules,
+  >;
+  constructor(queries: this['queries']) {
+    this.queries = queries;
+  }
+  toString(): string {
+    return this.queries.map((query) => query.toString()).join(' or ');
```

Comment:
Media query syntax doesn't actually allow for "or" keywords

```suggestion
    return this.queries.map((query) => query.toString()).join(', ');
```

Reply:
nvm, looking at the docs: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_media_queries/Using_media_queries

```
Note that you cannot use the or operator on the same level as the and and not operators. You can either separate the media features with a comma or use parenthesis to group sub-expressions of media features to clarify the order of evaluation.
```